---
title: "Final_VB_Assingment"
author: "Vishesh Bhatia"
date: "3/23/2021"
output: html_document
---

In the review part, Juan ended up making changes in my initial file and I was not able to retrieve the original version. I ended up accepting the pull request but I am making a new file with some more changes suggested by my peers and editing VB_Assignment.Rmd file. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
setwd('.')
```

## Reading the Genotypes and SNPs file
Through the code below, we are able to load the files fang_et_al_genotypes.txt 
and snp_position file into our global environment tab. 
```{r}
genotypes <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
SNPs <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/main/assignments/UNIX_Assignment/snp_position.txt", header = TRUE, sep = "\t")
```

# Part 1
## DATA INSPECTION

We were able to analyze both the genotypes and SNP files and the code for that is below. 

```{r}
dim(genotypes)
#Provided us with the dimensions of the file: 2782 rows and 986 columns
dim(SNPs)
#Provided us with the dimensions of the file: 983 rows and 15 columns

genotypes[1:10, 1:10]
SNPs[1:10, 1:10]
#Prints the first 10 rows and columns and helps us get an idea of what the file looks like. 

class(genotypes)
class(SNPs)
#Provides us the information about the vectors and looks like both files are data.frame format. 

str(genotypes)
str(SNPs)
#Provides information about structure of the data frame such as name, type and preview of data in each column. 

```
## PART 2
## Data Processing

Sort out the genotypes based on the Group and create separate data files of each species. 

```{r}
#Initially, I used the following code 
#filter (genotypes, Group == 'ZMMIL' | Group =='ZMMLR' | Group == 'ZMMMR') but Juan ended up changing to this. 
#Filter out the maize data
maize_g <- filter(genotypes, Group %in% c('ZMMIL', 'ZMMLR', 'ZMMMR'))

#Filter out the Teosinte Data
teosinte_g <- filter(genotypes, Group %in% c('ZMPBA', 'ZMPIL', 'ZMPJA'))
```

### Transposing the Maize Data

```{r}
maize_g <- column_to_rownames(maize_g, var = "Sample_ID")
transposed_maize <- t(maize_g) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID") 
transposed_maize <- transposed_maize[3:nrow(transposed_maize),]
Mdata <- merge(SNPs, transposed_maize, by = "SNP_ID")
Mdata <- select(Mdata, SNP_ID, Chromosome, Position, everything())
```

### Tranposing the Teosinte Data

```{r}
teosinte_g <- column_to_rownames(teosinte_g, var = "Sample_ID")
transposed_teosinte <- t(teosinte_g) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID") 
transposed_teosinte <- transposed_teosinte[3:nrow(transposed_teosinte),]
TData <- merge(SNPs, transposed_teosinte, by = "SNP_ID")
TData <- select(TData, SNP_ID, Chromosome, Position, everything())
```

## Maize Output

```{r warning=FALSE}
dir.create('./Maize') # ignore it if you have created this folder when creating the repository in github, 
chr_maize <- filter(Mdata, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(chr_maize$Chromosome))){
  chrm <-  chr_maize %>% filter(Chromosome == i) %>% arrange(Position)
  write.table(chrm, file = paste("./Maize/Maize_Part_A",i, sep="_"))
}

```

```{r}
Subbed_Maize <- as_tibble(lapply(Mdata, gsub, pattern ="?", replacement ="-", fixed = TRUE))
Subbed_Chr_Maize <- filter(Subbed_Maize, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(Subbed_Chr_Maize$Chromosome))) {
  chrm_subbed <- Subbed_Chr_Maize %>% filter(Chromosome == i) %>% arrange(desc(Position))
  write.table(chrm_subbed, file = paste("./Maize/Maize_Part_B", i, sep = "_"))
  }

```

## Teosinte Data
```{r warning=FALSE}
dir.create('./Teosinte') # same as maize output folder
chr_teosinte <- filter(TData, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(chr_teosinte$Chromosome))) {
  chrt <- chr_teosinte %>% filter(Chromosome == i) %>% arrange(Position)
  write.table (chrt, file = paste("./Teosinte/Teosinte_Part_A", i, sep = "_"))
}
```

```{r}
Subbed_Teosinte <- as_tibble(lapply(TData, gsub, pattern = "?", replacement = "-", fixed = TRUE))
Subbed_Chr_Teosinte <- filter(Subbed_Teosinte, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(Subbed_Chr_Teosinte$Chromosome))) {
    chrt_subbed <- Subbed_Chr_Teosinte %>% filter(Chromosome == i) %>% arrange(desc(Position))
    write.table(chrt_subbed, file = paste("./Teosinte/Teosinte_Part_B", i, sep = "_"))
}
```

# PART 3
# DATA VISUALIZATION

```{r}
# A little clean up on the original SNP and genotypes files. 
Formated_SNP <- SNPs %>% select(SNP_ID, Chromosome, Position)

Genotypes_Transposed <- genotypes %>% select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")


Merged_Genotypes <- merge(Formated_SNP, Genotypes_Transposed) %>% filter(., !Chromosome %in% c("unknown", "multiple"))

```

SNP DATA
```{r Modification to TotalSNPs and Diversity}
TotalSNPs <- 
  ggplot(Merged_Genotypes, aes(x=as.double(Chromosome), #x=as.double(Chromosome) considers Chr as a number
                               fill = as.factor(as.double(Chromosome)))) +  # for getting chrom in the right order in the legend   
    geom_bar() +
    scale_x_continuous(breaks = 1:10) + # bcs x=as.double(Chromosome), breaks by default don't match chrom number
    theme_bw() + 
   ggtitle("Total Number of SNPs across each chromosome") +
    labs(x = "Chromosome", y = "Total number of SNPs", fill = "Chromosome") #fill = "Chromosome" only replace legend's tittle 

DiversitySNPs <- 
  ggplot(Merged_Genotypes, aes(x= as.numeric(Position))) + 
  geom_density(aes(fill = as.factor(as.double(Chromosome)))) +  #same as the previous plot 
  facet_wrap(~ as.factor(as.double(Chromosome)), nrow = 2, ncol = 5) + 
  ggtitle("Diversity of SNPs across each chromosome") +
  theme(axis.text.x=element_text(angle = 90)) +                 #change orientation of x axis
  labs(x = "Position", y = "Density", fill = "Chromosome")

pdf("SNP_Visualisation.pdf")
print(TotalSNPs)
print(DiversitySNPs)
dev.off()
```

## Further Data Processing

```{r}
tidy_genotypes <- genotypes %>% select(-JG_OTU) %>% pivot_longer( -Sample_ID: -Group, names_to = "SNP_ID", values_to = "Sequence")
tidy_genotypes <- tidy_genotypes %>% mutate(new_sequence = ifelse(Sequence %in% c("A/A", "T/T", "C/C", "G/G"), "Homozygous", ifelse(Sequence == "?/?", "Missing", "Heterozygous")))
```


## Visualization of the missing data and amount of heterozygosity. 
```{r}
#For all the samples.
Samples_Plot <-  ggplot(tidy_genotypes, aes(x = Sample_ID, fill = new_sequence)) +
  ggtitle("Heterozygosity Plot") +
geom_bar(position = "fill") + theme_bw() + labs(x = "Sample ID", y = "Proportion")

#Stacked Bar-graph for all groups. 
Groups_Plot <- ggplot(tidy_genotypes, aes(x = Group , fill = new_sequence)) + geom_bar(position = "fill") + 
  ggtitle("Heterozygosity in various Groups of Species ") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90))+ labs(y = "Proportion")

pdf("MissingDATA_Heterozygosity_Visualisation.pdf")
print(Samples_Plot)
print(Groups_Plot)
dev.off()

```

##  My own visualization

```{r}
OwnFeature <- ggplot(filter(tidy_genotypes, Sequence != "?/?") , aes(x = Sample_ID, fill = Sequence)) + 
  geom_bar(position = "fill") + theme_bw() + labs(x = "Sample ID", y = "Proportion")

pdf("Own_Feature.pdf")
print(OwnFeature)
dev.off()
```

